api/src/main/java/com/example/api/util/visitor/HeroVisitor.java

```diff
@@ -0,0 +1,79 @@
+package com.example.api.util.visitor;
+
+import com.example.api.dto.response.activity.result.SuperPowerResponse;
+import com.example.api.error.exception.RequestValidationException;
+import com.example.api.model.activity.result.GraphTaskResult;
+import com.example.api.model.activity.result.ResultStatus;
+import com.example.api.model.question.Question;
+import com.example.api.model.question.QuestionType;
+import com.example.api.model.user.User;
+import com.example.api.model.user.hero.*;
+import com.example.api.util.calculator.TimeCalculator;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class HeroVisitor {
+    private final TimeCalculator timeCalculator;
+    public SuperPowerResponse<Long> visitPriest(Priest priest,
+                                             User user,
+                                             GraphTaskResult result) throws RequestValidationException {
+        checkIfHeroPowerCanBeUsed(priest, user, result);
+
+        int level = user.getLevel();
+        long healValue = priest.getHealValueForLevel(level);
+
+        long startDateMillis = result.getStartDateMillis();
+        long newStartDateMillis = startDateMillis + healValue;
+        long timeToSolveMillis = result.getGraphTask().getTimeToSolveMillis();
+
+        result.setStartDateMillis(newStartDateMillis);
+        user.getUserHero().setLastSuperPowerUsageTimeMillis(System.currentTimeMillis());
+        return new SuperPowerResponse<>(timeCalculator.getTimeRemaining(newStartDateMillis, timeToSolveMillis));
+    }
+
+    public SuperPowerResponse<Boolean> visitRogue(Rogue rogue,
+                                            User user,
+                                            GraphTaskResult result) throws RequestValidationException {
+        checkIfHeroPowerCanBeUsed(rogue, user, result);
+        result.setStatus(ResultStatus.CHOOSE);
+
+        user.getUserHero().setLastSuperPowerUsageTimeMillis(System.currentTimeMillis());
+        return new SuperPowerResponse<>(true);
+    }
```

Comment:
Zastanawia mnie umiejetność ROUGE, bo jeśli dobrze zrozumiałem to wykonując jego umiejętność pomijania nie dodajemy żadnej odpowiedzi jak przy normalnym flow, czyli nie dostajemy też żadnych punktów. Ale wtedy chyba nie ma sensu korzystać z tego no bo lepiej strzelić odpowiedź i mieć szanse na dostanie jakichś punktów.

Reply:
Różnica jest dość prosta - jak jej użyjesz to nie liczysz punktów z pytania w mianowniku. A jak strzelisz i nie trafisz to masz np 0/10

Czyli np mając 4 pytania w ekspedycji, każde po 10pkt to gdy nie użyjesz mocy i strzelisz w 0 a pozostałe w maxa to masz na koniec 30/40. A jak użyjesz mocy i w tym co dostałeś zero byś pominął to byś miał 30/30. Mam nadzieję, że to zrozumiałe. A przynajmniej taka była idea, mam nadzieję że tak to @ajablonsk1 zrobił 😅

Reply:
Tak to dokladnie dziala

Reply:
Rozumiem, tylko w  tym pr Adrian wywalił max100, który nie był już używany nigdzie, a który pierwotnie służył chyba do liczenia mianownika (customowego dla wybranej ścieżki)

Reply:
Nope, maxPoints100 nie bylo nigdzie uzywane, dlatego wywalilem

Reply:
Punkty max za sciezke jaka szedl licza sie tak, ze w graph result sa answers i z nich brane jest getQuestion.getPoints() i sumowane, jak rogue zrobi umiejetnosc to zadna odp sie nie dodaje do resulta

Reply:
A dobra, widzę już gdzie to jest liczone